### ISSUE TITLE
R script unable to access model

### ISSUE THREAD

**Issue by samkleeman1** [ncborcherding/Ibex#3]
Hi,

Many thanks for making this very interesting package available.

I am currently following the tutorial using some data we have collected internally.

I am seeing an error message when I run:

`SeuratObj <- runIbex(SeuratObj, 
                     chains = "Heavy",
                     AA.properties = "AF", 
                     reduction.name = "ibex.KF")`

Error is:

`[1] "Calculating the Amino Acid Properties..."
Invalid file name (invalid file name)Traceback:

1. runIbex(SeuratObj, chains = "Heavy", AA.properties = "AF", reduction.name = "ibex.KF")
2. Ibex.matrix(sc, chains, AA.properties)
3. aaProperty(BCR, AA.properties)
4. quiet(aa.model.loader(chain[[i]], AA.properties))
5. force(x)
6. aa.model.loader(chain[[i]], AA.properties)
7. quiet(load_model_hdf5(select, compile = FALSE))
8. force(x)
9. load_model_hdf5(select, compile = FALSE)
10. load_model(filepath, custom_objects, compile)
11. do.call(keras$models$load_model, args)
12. (structure(function (filepath, custom_objects = NULL, compile = TRUE) 
  . {
  .     cl <- sys.call()
  .     cl[[1L]] <- list2
  .     call_args <- split_named_unnamed(eval(cl, parent.frame()))
  .     result <- py_call_impl(callable, call_args$unnamed, call_args$named)
  .     if (convert) 
  .         result <- py_to_r(result)
  .     if (is.null(result)) 
  .         invisible(result)
  .     else result
  . }, class = c("python.builtin.function", "python.builtin.object"
  . ), py_object = <environment>))(filepath = "", custom_objects = NULL, 
  .     compile = FALSE)
13. py_call_impl(callable, call_args$unnamed, call_args$named)`

The underlying command `select  <- system.file("extdata", paste0(chain, "_", AA.properties, "_Encoder.h5"), package = "Ibex")` does not seem to link to anything, unsure why.

Many thanks in advance!

BW, Sam Kleeman, Cold Spring Harbor Laboratory

**Comment by samkleeman1** [comment]
This is both on the DEV and MASTER forks

**Comment by ncborcherding** [comment]
Hey Sam, 

Thanks for reaching out and happy to help - could you give me the output of ```sessionInfo()```?  

You might also want to check that your library paths are corresponding to where Ibex was installed by comparing:

```
.libPaths()
system.file("extdata", "heavy_AF_Encoder.h5", package="Ibex")
```
If truly not there - the system.fil() call should return an NA or error. Not sure if it is a version issue - the current version is 0.99.3 and there is no dev fork in Ibex (as of yet). 

Thanks,
Nick

**Comment by samkleeman1** [comment]
Many thanks for the rapid response

I think I see the issue - it relates to capitalization of the word 'Heavy'. If I use lower case 'heavy' it can get past that step, then I get a new error due to the code (https://github.com/ncborcherding/Ibex/blob/5d64fb7787b25eb53307d62e1224eb69248ffed6/R/utils.R#L23).

New error is:
`Error in getBCR(sc, chains): object 'pos' not found
Traceback:

1. runIbex(SeuratObj, chains = "heavy", AA.properties = "KF", reduction.name = "ibex.KF")
2. Ibex.matrix(sc, chains, AA.properties)
3. getBCR(sc, chains)
`

**Comment by ncborcherding** [comment]
Hey Sam, 

Most recent commit (v0.99.4) added the internal function chain.checker() that prevents this issue. Thanks for pointing it out and let me know if you have any suggestions.

Nick

**Comment by samkleeman1** [comment]
Sorry to write back.

Once I get past this error I then get a different one:

`[1] "Calculating the Amino Acid Properties..."
Invalid file name (invalid file name)Traceback:

1. runIbex(SeuratObj, chains = "Heavy", AA.properties = "KF", reduction.name = "ibex.KF")
2. Ibex.matrix(sc, chains, AA.properties)
3. aaProperty(BCR, AA.properties)
4. quiet(aa.model.loader(chain[[i]], AA.properties))
5. force(x)
6. aa.model.loader(chain[[i]], AA.properties)
7. quiet(load_model_hdf5(select, compile = FALSE))
8. force(x)
9. load_model_hdf5(select, compile = FALSE)
10. load_model(filepath, custom_objects, compile)
11. do.call(keras$models$load_model, args)
12. (structure(function (filepath, custom_objects = NULL, compile = TRUE) 
  . {
  .     cl <- sys.call()
  .     cl[[1L]] <- list2
  .     call_args <- split_named_unnamed(eval(cl, parent.frame()))
  .     result <- py_call_impl(callable, call_args$unnamed, call_args$named)
  .     if (convert) 
  .         result <- py_to_r(result)
  .     if (is.null(result)) 
  .         invisible(result)
  .     else result
  . }, class = c("python.builtin.function", "python.builtin.object"
  . ), py_object = <environment>))(filepath = "", custom_objects = NULL, 
  .     compile = FALSE)
13. py_call_impl(callable, call_args$unnamed, call_args$named)
`

**Comment by ncborcherding** [comment]
Hey Sam, 

Did you restart your rsession after installing the update?

Nick

**Comment by samkleeman1** [comment]
Yes

**Comment by samkleeman1** [comment]
The script I am running is as follows (it is a dataset of ours)

`

suppressMessages(library(Ibex))
suppressMessages(library(Seurat))
suppressMessages(library(ggplot2))
suppressMessages(library(viridis))
suppressMessages(library(patchwork))

suppressMessages(library(scRepertoire))
##################################
#scRNA/ADT loading and processing
#################################
tmp <-  Read10X("/mnt/grid/janowitz/home/data/4t1_nmdar_tumorseq/april23/4t1_cohortb_pooled_correctseq_nrref/outs/per_sample_outs/4t1_cohortb_pooled_correctseq_nrref/count/sample_filtered_feature_bc_matrix")

MIS.sample <- CreateSeuratObject(counts = tmp$`Gene Expression`)

adt_assay <- CreateAssayObject(counts = tmp$`Antigen Capture`)

MIS.sample[["ADT"]] <- adt_assay
MIS.sample <- subset(MIS.sample, subset = nFeature_RNA > 200) 
MIS.sample  <- RenameCells(object = MIS.sample , new.names = paste0("MIS.sample_", rownames(MIS.sample[[]])))
MIS.sample[["mito.genes"]] <- PercentageFeatureSet(MIS.sample, pattern = "^MT-")
    
#Filtering step
standev <- sd(log(MIS.sample$nFeature_RNA))*2.5 #cutting off above standard deviation of 2.5
mean <- mean(log(MIS.sample$nFeature_RNA))
cut <- round(exp(standev+mean))
MIS.sample <- subset(MIS.sample, subset = mito.genes < 10 & nFeature_RNA < cut)

#Processing and Adding Contig Info
contigs <- read.csv("/mnt/grid/janowitz/home/data/4t1_nmdar_tumorseq/april23/4t1_cohortb_pooled_correctseq_nrref/outs/per_sample_outs/4t1_cohortb_pooled_correctseq_nrref/vdj_b/filtered_contig_annotations.csv")
clones <- combineBCR(list(contigs), samples = "MIS.sample",removeNA = TRUE)
MIS.sample <- combineExpression(clones, MIS.sample, cloneCall="aa")

library(dplyr)
#Subset only B cells (by contigs)
MIS.sample$BCR.recoverd <- "No"
MIS.sample$BCR.recoverd[!is.na(MIS.sample$CTaa)] <- "Yes"
MIS.sample <- subset(MIS.sample, BCR.recoverd == "Yes")

#Processing RNA
DefaultAssay(MIS.sample) <- 'RNA'
MIS.sample <- NormalizeData(MIS.sample) %>% FindVariableFeatures() %>% 
  quietBCRgenes() %>% ScaleData() %>% RunPCA(verbose = FALSE)

SeuratObj=MIS.sample

SeuratObj <- runIbex(SeuratObj, 
                     chains = "Heavy",
                     AA.properties = "OHE", 
                     reduction.name = "ibex.KF")


`


The error message is as follows:

`
[1] "Calculating the Amino Acid Properties..."
'str' object has no attribute 'decode'Traceback:

1. runIbex(SeuratObj, chains = "Heavy", AA.properties = "OHE", reduction.name = "ibex.KF")
2. Ibex.matrix(sc, chains, AA.properties)
3. aaProperty(BCR, AA.properties)
4. quiet(aa.model.loader(chain[[i]], AA.properties))
5. force(x)
6. aa.model.loader(chain[[i]], AA.properties)
7. quiet(load_model_hdf5(select, compile = FALSE))
8. force(x)
9. load_model_hdf5(select, compile = FALSE)
10. load_model(filepath, custom_objects, compile)
11. do.call(keras$models$load_model, args)
12. (structure(function (filepath, custom_objects = NULL, compile = TRUE) 
  . {
  .     cl <- sys.call()
  .     cl[[1L]] <- list2
  .     call_args <- split_named_unnamed(eval(cl, parent.frame()))
  .     result <- py_call_impl(callable, call_args$unnamed, call_args$named)
  .     if (convert) 
  .         result <- py_to_r(result)
  .     if (is.null(result)) 
  .         invisible(result)
  .     else result
  . }, class = c("python.builtin.function", "python.builtin.object"
  . ), py_object = <environment>))(filepath = "/grid/wsbs/home_norepl/skleeman/R/x86_64-pc-linux-gnu-library/4.0/Ibex/extdata/Heavy_OHE_Encoder.h5", 
  .     custom_objects = NULL, compile = FALSE)
13. py_call_impl(callable, call_args$unnamed, call_args$named)


`


sessionInfo() is as follows:


`

R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Rocky Linux 8.7 (Green Obsidian)

Matrix products: default
BLAS/LAPACK: /grid/it/data/elzar/easybuild/software/OpenBLAS/0.3.9-GCC-9.3.0/lib/libopenblas_skylakexp-r0.3.9.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_1.1.2         scRepertoire_1.11.0 patchwork_1.1.2    
[4] viridis_0.6.3       viridisLite_0.4.2   ggplot2_3.4.2      
[7] SeuratObject_4.1.3  Seurat_4.3.0        Ibex_0.99.4        

loaded via a namespace (and not attached):
  [1] uuid_1.1-0                  VGAM_1.1-5                 
  [3] plyr_1.8.8                  igraph_1.5.0               
  [5] repr_1.1.3                  lazyeval_0.2.2             
  [7] sp_2.0-0                    splines_4.0.3              
  [9] powerTCR_1.10.3             BiocParallel_1.24.1        
 [11] listenv_0.9.0               scattermore_1.2            
 [13] tfruns_1.5.1                GenomeInfoDb_1.26.7        
 [15] digest_0.6.33               foreach_1.5.2              
 [17] htmltools_0.5.5             ggalluvial_0.12.3          
 [19] fansi_1.0.4                 truncdist_1.0-2            
 [21] magrittr_2.0.3              tensor_1.5                 
 [23] cluster_2.1.2               doParallel_1.0.17          
 [25] ROCR_1.0-11                 graphlayouts_1.0.0         
 [27] globals_0.16.2              matrixStats_1.0.0          
 [29] R.utils_2.12.2              spatstat.sparse_3.0-2      
 [31] colorspace_2.1-0            ggrepel_0.9.3              
 [33] crayon_1.5.2                RCurl_1.98-1.12            
 [35] jsonlite_1.8.7              progressr_0.13.0           
 [37] spatstat.data_3.0-1         zeallot_0.1.0              
 [39] iterators_1.0.14            survival_3.2-11            
 [41] zoo_1.8-12                  glue_1.6.2                 
 [43] polyclip_1.10-4             gtable_0.3.3               
 [45] zlibbioc_1.36.0             XVector_0.30.0             
 [47] leiden_0.4.3                DelayedArray_0.16.3        
 [49] evd_2.3-6.1                 future.apply_1.11.0        
 [51] SingleCellExperiment_1.12.0 BiocGenerics_0.47.0        
 [53] SparseM_1.81                abind_1.4-5                
 [55] scales_1.2.1                spatstat.random_3.1-5      
 [57] miniUI_0.1.1.1              Rcpp_1.0.11                
 [59] xtable_1.8-4                reticulate_1.30            
 [61] stats4_4.0.3                htmlwidgets_1.6.2          
 [63] httr_1.4.6                  RColorBrewer_1.1-3         
 [65] ellipsis_0.3.2              ica_1.0-3                  
 [67] R.methodsS3_1.8.2           farver_2.1.1               
 [69] pkgconfig_2.0.3             uwot_0.1.16                
 [71] deldir_1.0-9                utf8_1.2.3                 
 [73] tidyselect_1.2.0            rlang_1.1.1.9000           
 [75] reshape2_1.4.4              later_1.3.1                
 [77] munsell_0.5.0               tools_4.0.3                
 [79] cli_3.6.1                   generics_0.1.3             
 [81] ggridges_0.5.4              evaluate_0.21              
 [83] stringr_1.5.0               fastmap_1.1.1              
 [85] goftest_1.2-3               evmix_2.12                 
 [87] tidygraph_1.2.3             fitdistrplus_1.1-11        
 [89] purrr_1.0.1                 RANN_2.6.1                 
 [91] ggraph_2.1.0                pbapply_1.7-2              
 [93] future_1.33.0               nlme_3.1-152               
 [95] whisker_0.4.1               mime_0.12                  
 [97] R.oo_1.25.0                 compiler_4.0.3             
 [99] plotly_4.10.2               png_0.1-8                  
[101] spatstat.utils_3.0-3        tweenr_2.0.2               
[103] tibble_3.2.1                gsl_2.1-6                  
[105] stringi_1.7.12              cubature_2.0.4.2           
[107] lattice_0.20-44             bluster_1.2.1              
[109] IRdisplay_1.0               Matrix_1.6-0               
[111] tensorflow_2.11.0           keras_2.11.1               
[113] permute_0.9-5               vegan_2.5-7                
[115] vctrs_0.6.3                 stringdist_0.9.10          
[117] pillar_1.9.0                lifecycle_1.0.3            
[119] spatstat.geom_3.2-2         lmtest_0.9-40              
[121] RcppAnnoy_0.0.21            BiocNeighbors_1.8.2        
[123] data.table_1.14.8           cowplot_1.1.1              
[125] bitops_1.0-7                irlba_2.3.5.1              
[127] httpuv_1.6.11               GenomicRanges_1.42.0       
[129] R6_2.5.1                    promises_1.2.0.1           
[131] KernSmooth_2.23-20          gridExtra_2.3              
[133] IRanges_2.24.1              parallelly_1.36.0          
[135] codetools_0.2-18            MASS_7.3-55                
[137] SummarizedExperiment_1.20.0 withr_2.5.0                
[139] sctransform_0.3.5           S4Vectors_0.39.1           
[141] GenomeInfoDbData_1.2.4      mgcv_1.8-35                
[143] parallel_4.0.3              grid_4.0.3                 
[145] IRkernel_1.2                tidyr_1.3.0                
[147] MatrixGenerics_1.2.1        Rtsne_0.16                 
[149] ggforce_0.4.1               pbdZMQ_0.3-5               
[151] spatstat.explore_3.2-1      Biobase_2.50.0             
[153] shiny_1.7.4.1               base64enc_0.1-3            


`

**Comment by ncborcherding** [comment]
Everything looks good to go from your code end and from your sessioinfo. 

This is likely an issue with the underlying python versioning. It looks like it is locating the model in your R library, but is not able to read? 

How did you set up your environment for reticulate? Are you using conda? If so, trying a virtualenv for this might be easier.

Nick

**Comment by samkleeman1** [comment]
Hi Nick, thanks so much for your help with this. I was able to resolve it by re-installing reticulate/keras/TF according to the following instructions (may be helpful for others)

https://hastie.su.domains/ISLR2/keras-instructions.html

Cool package!

**Comment by samkleeman1** [comment]
Actually - sorry to say I reached a new error. It did run correctly (once), but now I get the following error persistently. I tried reinstalling the package.

```
[1] "Calculating the Amino Acid Properties..."
SavedModel file does not exist at: /{saved_model.pbtxt|saved_model.pb}Traceback:

1. runIbex(SeuratObj, chains = "Heavy", AA.properties = "AF", reduction.name = "ibex.KF")
2. Ibex.matrix(sc, chains, AA.properties)
3. aaProperty(BCR, AA.properties)
4. quiet(aa.model.loader(chain[[i]], AA.properties))
5. force(x)
6. aa.model.loader(chain[[i]], AA.properties)
7. quiet(load_model_hdf5(select, compile = FALSE))
8. force(x)
9. load_model_hdf5(select, compile = FALSE)
10. load_model(filepath, custom_objects, compile)
11. do.call(keras$models$load_model, args)
12. (structure(function (filepath, custom_objects = NULL, compile = TRUE, 
  .     options = NULL) 
  . {
  .     cl <- sys.call()
  .     cl[[1L]] <- list2
  .     call_args <- split_named_unnamed(eval(cl, parent.frame()))
  .     result <- py_call_impl(callable, call_args$unnamed, call_args$named)
  .     if (convert) 
  .         result <- py_to_r(result)
  .     if (is.null(result)) 
  .         invisible(result)
  .     else result
  . }, class = c("python.builtin.function", "python.builtin.object"
  . ), py_object = <environment>))(filepath = "", custom_objects = NULL, 
  .     compile = FALSE)
13. py_call_impl(callable, call_args$unnamed, call_args$named)
```

**Comment by ncborcherding** [comment]
Apologies for the delay - I think what must be going on here is there is an issue with the way your python environment is linked - especially if 1) was able to run at least once and 2) the .h5 model files can be located in the R library. I'll have to investigate more, but thanks for bearing with me. 

Nick

**Comment by EvgenEgorov** [comment]
Hello!
This question is not related to this particular issue - is it possible to run Ibex.matrix or runIbex on both chains simultaneously? Or is it always necessary to choose only one chain (heavy or light)?

**Comment by ncborcherding** [comment]
As of right now – Ibex works on single-chains at a time. This was to allow for the collection of the most number of sequences possible for training. The eventual goal is to have dual chain embeddings as well. 

In the meantime, you can run the chains individually and use a multimodal integration method. For example, on the TCR side of things (using the Trex package), I have had a lot of luck running TCRA and TCRB and then combining them into a single representation using fastmnn (https://bioconductor.org/packages/release/bioc/html/batchelor.html). I am still exploring the use WNN or the scvi framework.
